### DISCUSSION TITLE
Simple example of iterating over struct lyd_node

### DISCUSSION THREAD

**Discussion by yogch** [CESNET/libyang#1690] [Q&A]
I'm trying to parse a response I get using libnetconf2 and libyang. I could not find any simple example that's doing so.  
I'm using `nc_rpc_get()`. Then I get an `rpc` and use `nc_send_rpc()`. Then I fetch the data using `nc_recv_reply()`.  
`nc_recv_reply()` gives me a `libyang` object which is called `struct lyd_node*` which should reflect (AFAIU) the data in yang.  
The problem is that it just doesn't bring the data in data structures. I can only print it with `lyd_any_value_str()` and I get an `xml`. I can't really go over the nodes and see what is a `container`, what is a `leaf` and get that value.  

What I'm asking is for a simple example of querying a server and then iterating over the results as if one node was `list`, then another was a `leaf` etc...

Thanks anyway!

**Comment by michalvasko** [comment]
Okay, so you get an `anyxml` node (which you can learn by examining `struct lyd_node *node->schema->nodetype`). If you cast it properly into `struct lyd_node_any *`, you will get 2 new members `value` and `value_type`. What you should get in *libnetconf2* responses is `value_type` of `LYD_ANYDATA_DATATREE`. That means you can access `value.tree`, which is probably the data tree you are looking for.

**Reply by yogch** [reply]
Hi,
Thank you for the reply.

My query is a simple `get`: `/cont-a/cont-b/list-a`.
What I get as the returned `lyd_node` from `nc_recv_reply()` is:
- schema->name: "get"
- schema->nodetype: 256.

I cast and get `tree`. Its schema->name is "data" and the type is 32.
Afterwards there's no more data nodes.
I mean, if I use `lyd_child()` and `next`. There are no more nodes inside.
If I use `lyd_any_value_str()` though, it prints the full returned XML.

I was able to get the data using `(struct lyd_node_opaq*)node)->child` casting.
The thing is, that's `lyd_node_opaq` (or `lyd_node_inner`).
And it says: "Data node structure for unparsed (opaque) nodes."

And I do not understand this one - I thought I should get a parsed node which says what's the type of the node (container / list / leaf).
Instead I get unparsed nodes and iterate over them like a simple xml.

Is this the expected result?
Thanks!

**Reply by michalvasko** [reply]
Yes, this all is expected if you simply look at the [YANG module](https://github.com/CESNET/netopeer2/blob/devel/modules/ietf-netconf%402013-09-29.yang#L757). And I have answered exactly what you are asking about, again, so please reread what I wrote, not sure what else to add.

**Reply by yogch** [reply]
Ok,
I thought `libyang` automatically converts the returned `xml` from `libnetconf2` into `struct lyd_node` using the schema.
So I can iterate over it.
For my example, that it will contain a `schema` with `nodestype`: container, container, list. For my `/cont-a/cont-b/list-a`.
And that I'll be able to fetch the value using `lyd_get_value()`.

Thanks.

**Reply by michalvasko** [reply]
*libyang* does automatically convert the returned XML into `struct lyd_node *` data tree, you just have to get it from the `struct lyd_node_any *` structure first, which I have described how to do still in my first post. So yet again, please reread it.

**Reply by yogch** [reply]
Hmm,
I do get the `tree`:
```
struct lyd_node_any* any = (struct lyd_node_any*)lyd_node;
struct lyd_node* tree = any->value.tree;
```

Now, this is the tree values:
![image](https://user-images.githubusercontent.com/13855845/131508322-d7ee2562-9bf2-456b-b292-b55aea2fbd89.png)

And here, if I cast it to `struct lyd_node_inner*`:
![image](https://user-images.githubusercontent.com/13855845/131508420-3fb19af4-ffdf-4648-b19a-e99feb35e709.png)

The way I'm currently able to get values is:
```
struct lyd_node* list = ((struct lyd_node_opaq*)((struct lyd_node_opaq*)((struct lyd_node_opaq*)((struct lyd_node_opaq*)node)->child)->child)->child)->child;   // list of elements head - 'cont-a/cont-b/list-a
```
Then I do again: 
```
struct lyd_node* list_entry = ((struct lyd_node_opaq*)list)->child
```
In order to get a single list entry.
Then I'm able to iterate over the `list_entry->next` to go over all leafs inside.

Does it make sense ?

**Reply by michalvasko** [reply]
No, you first need to get the child of the RPC node, simple `lyd_child()`, that should get you the `data` node (`name` in its `schema`) first. This is the `anyxml` node that needs to be cast to `struct lyd_node_any *` in which you can get the data `tree` and then to iterate over that one is what you want. Usually with macros like `LY_LIST_FOR` or `LYD_TREE_DFS_BEGIN/END`.

**Reply by yogch** [reply]
Doesn't seem it helps :/
```
struct lyd_node_any* any = (struct lyd_node_any*)lyd_child(lyd_node);
struct lyd_node* tree = any->value.tree;
```

![image](https://user-images.githubusercontent.com/13855845/131560377-460e53b0-1b5b-4bdb-aa1d-e2ba58e757af.png)

**Reply by michalvasko** [reply]
Look, it is up to you to cast everything properly, I do not know what exactly you have in `lyd_node`. In general, look at `schema->nodetype` in every `struct lyd_node *`. Based on the value there, you cast it to a specific node. If `nodetype & LYD_NODE_INNER`, you cast it to `struct lyd_node_inner *`, if `nodetype & LYD_NODE_TERM`, use `struct lyd_node_term *`, and if `nodetype & LYD_NODE_ANY`, use `struct lyd_node_any *`. I cannot help you more.

**Reply by yogch** [reply]
The thing is that `schema` is NULL, so I can't look at `schema->nodetype` for any node after `data` (`next` or `child`).
It treats all values as `struct lyd_node_opaq`.
Thanks anyway.

**Reply by michalvasko** [reply]
Oh, I see, somehow missed that, sorry. You can enable *libyang* verbose output and it should tell you when it is parsing opaque nodes. But in this case the reason will most likely be that you do not have the YANG modules in the context for the data you have received.

**Reply by yogch** [reply]
You mean using `nc_client_set_schema_searchpath()` ?
I'm using that. There are couple of mandatory .yang files that must be there, without it the netconf request doesn't work at all.
Then I've tried putting my schema in the searchpath but nothing changes :/

**Reply by michalvasko** [reply]
Try to increase verbosity using `nc_verbosity()` to INF and you should see how the schemas are loaded and whether any fails to.

**Reply by yogch** [reply]
Looks like lots of schemas fail :/
I'll try to check it out..

**Reply by yogch** [reply]
Well, I have no clue what's happening. :/

```
[INF]: Session 157: retrieving data for schema "ddd", revision "2016-07-10".
[INF]: Session 157: reading schema from server via get-schema.
[WRN]: Failed to load schema "ddd@2016-07-10".
```

This is specifically a very simple schema:
```
module ddd {
  namespace "http://ddd.com";
  prefix ddd;

  organization "DDD";
  description
    "This module is the parent module";

  revision 2016-07-10 {
    description
      "Initial release";
  }
}
```

There are other schema-load failures as well

**Reply by michalvasko** [reply]
It is strange that you do not see any other message so it is suspicious, *libyang* loads this module just fine. What versions are you using?

**Reply by yogch** [reply]
`v2.0.7`

**Reply by michalvasko** [reply]
For a quick fix, try using the latest v2 in the `devel` branch. It should be released officially soon.
